core/src/main/java/org/apache/iceberg/IncrementalFileCleanup.java

```diff
@@ -0,0 +1,349 @@
+        .executeWith(deleteExecutorService)
+        .retry(3)
+        .stopRetryOn(NotFoundException.class)
+        .suppressFailureWhenFinished()
+        .onFailure((file, exc) -> LOG.warn("Delete failed for data file: {}", file, exc))
+        .run(deleteFunc::accept);
+  }
+
+  private Set<String> findFilesToDelete(
+      Set<ManifestFile> manifestsToScan, Set<ManifestFile> manifestsToRevert, Set<Long> validIds) {
+    Set<String> filesToDelete = ConcurrentHashMap.newKeySet();
+    Tasks.foreach(manifestsToScan)
+        .retry(3)
+        .suppressFailureWhenFinished()
+        .executeWith(planExecutorService)
+        .onFailure(
+            (item, exc) ->
+                LOG.warn("Failed to get deleted files: this may cause orphaned data files", exc))
+        .run(
+            manifest -> {
+              // the manifest has deletes, scan it to find files to delete
+              try (ManifestReader<?> reader =
+                  ManifestFiles.open(manifest, ops.io(), ops.current().specsById())) {
+                for (ManifestEntry<?> entry : reader.entries()) {
+                  // if the snapshot ID of the DELETE entry is no longer valid, the data can be
+                  // deleted
+                  if (entry.status() == ManifestEntry.Status.DELETED
+                      && !validIds.contains(entry.snapshotId())) {
+                    // use toString to ensure the path will not change (Utf8 is reused)
+                    filesToDelete.add(entry.file().path().toString());
+                  }
+                }
+              } catch (IOException e) {
+                throw new RuntimeIOException(e, "Failed to read manifest file: %s", manifest);
+              }
+            });
+
+    Tasks.foreach(manifestsToRevert)
+        .retry(3)
+        .suppressFailureWhenFinished()
+        .executeWith(planExecutorService)
+        .onFailure(
+            (item, exc) ->
+                LOG.warn("Failed to get added files: this may cause orphaned data files", exc))
+        .run(
+            manifest -> {
+              // the manifest has deletes, scan it to find files to delete
+              try (ManifestReader<?> reader =
+                  ManifestFiles.open(manifest, ops.io(), ops.current().specsById())) {
+                for (ManifestEntry<?> entry : reader.entries()) {
+                  // delete any ADDED file from manifests that were reverted
+                  if (entry.status() == ManifestEntry.Status.ADDED) {
+                    // use toString to ensure the path will not change (Utf8 is reused)
+                    filesToDelete.add(entry.file().path().toString());
+                  }
+                }
+              } catch (IOException e) {
+                throw new RuntimeIOException(e, "Failed to read manifest file: %s", manifest);
+              }
+            });
+
+    return filesToDelete;
+  }
+
+  private CloseableIterable<ManifestFile> readManifestFiles(Snapshot snapshot) {
```

Comment:
Why is this here and not in the base class?

Reply:
Oversight, moved